### ISSUE TITLE
Keep TabContent when switching tabs

### ISSUE THREAD

**Issue by nabinpuri** [valor-software/ng2-bootstrap#6060]
Currently, I see no support to preserve the tab content. Tab contents are created every time each tab becomes active. 

It would be nice to have new attribute/setting to disable/enable keeping content even when the tab is no more active. It seems ng-bootstrap support this out of the box with [destroyonhide]. Can we have similar support.

Also, can someone suggest the workaround for now?

**Comment by Jakartoine** [comment]
Hey @nabinpuri,

Could you please explain what is the purpose?

By default the content of a tab is "static", it means that even if a tab is not selected, its content is loaded and visible in the DOM. TabModule only provides (natively) an engine about showing/hiding content by tab clicking.

This, let me think that what you want is a way to keep a content displayed even if the tab where it is located is not active,
is that correct?

Feel free to post your code or be more explicit and I'll do my best to help you.

Cheers

**Comment by nabinpuri** [comment]
Hi,
Lets say, If you click on any tab, will the previous tab content(if loaded dynamically) fully destroyed or just hidden? Because when I see the documentation, It says whenever tab is selected the previous tab content(component) is destroyed. See below.

Dynamic content rendering
"The Component inside the Tab is rendered, when the tab is activated and destroyed when the tab is left".

I do not mean to keep the content displayed even if the tab is inactive. what I want is for the dynamic content to be remembered and no need to recreate again when tab is selected again.

Eg: If I have a table in first tab. The user apply some filter and the content is loaded with filtered data. Now, If user switch to different tabs, and come back the component inside the tab will recreate and the filtered applied will be lost.

Correct me If I am missing something.

Thanks.

**Comment by Jakartoine** [comment]
@nabinpuri,

Alright, I now understand your problem a little better.

As I said, the native behavior of TabModule is only to show/hide content based on tab clicking. That means that natively and basically, TabModule will not interfere with Angular lifecycle hook of your different components inside a tab.

The section you mentioned above is to make TabModule tab content dynamic!

So the only thing you have to do is to use the basic sample of TabModule for your case ;-)

Something like that:

![image](https://user-images.githubusercontent.com/27675999/113879717-1a3fcf00-97bb-11eb-9a3c-05d9231284cf.png)

Let me know if it works for you!

Cheers

**Comment by nabinpuri** [comment]
As per your comment, The tab content will be remembered even when the tab is inactive. If this the default behavior, then It should work for me. I was confused by the "dynamic content demo" [here ](https://valor-software.com/ngx-bootstrap/#/tabs).

It says.
The Component inside the Tab is rendered, when the tab is activated and **destroyed when the tab is left.**

Sub-Component A activated
Sub-Component A destroyed
Sub-Component B activated

` <tab heading="Tab A" #tabA="tab">
    <sub-component
      name="A"
      *ngIf="tabA.active"
      (onInit)="message('Sub-Component A activated')"
      (onDestroy)="message('Sub-Component A destroyed')"
    ></sub-component>
  </tab>`

`<tab heading="Tab B" #tabB="tab">
    <sub-component
      name="B"
      *ngIf="tabB.active"
      (onInit)="message('Sub-Component B activated')"
      (onDestroy)="message('Sub-Component B destroyed')"
    ></sub-component>
  </tab>`

 what does above demo and code imply? I understand that subcomponent is destroyed whenever tab is left. 

Anyway, from what you explain, I should be able to achieve what I want. In this case, If it does not work it should be bug rather than enhancement? But again, the demo is contradicting what you are saying.

Thank you

**Comment by Jakartoine** [comment]
@nabinpuri,

Yeah, it will work in your case, just use my sample.

There is no bug, this is as it.

**I think you can consider this as closed.**

The code you've shown above is exactly what you have understood, because they've added a structural directive on the active state of each tab :-) (which, in facts, add the dynamic part effect you have mentioned)

**FYI**

To go deeper, I had a use case in my code where I needed to have the tab content created only when a tab is active (because of huge data loaded in it) but not destroyed when the tab is unselected.

To achieve this, I reused their sample of dynamic part (as your code above) but I replaced de *ngIf="tabXxx.active" by
a custom structural directive which acts as an ngIf but only once : *ngIfOnce="tabXxx.active".

So using that, on the sub-component at page refreshing, only the tab content of the active tab is loaded. Others tab contents are not instanciated (increasing performance of my app). They will only be loaded if I click on their tab header.
Afterwards, if you switch between tabs they will use the native TabModule engine of only showing/hiding their tab content (without destroying it then).

--------------------------------------------------------------------------------------------------------------------------

So I agree that their samples are not fully explained but if you know how Angular works properly, you should succed to use their library without any big challenge ;-)

Cheers buddy!
